Fix book proficiency mitigation formulas.

eltank opened the pull request:
#### Summary
Bugfixes "Partially fix proficiency books' effect on crafting (formulas and UI)"

#### Purpose of change

This is a fork of #48650 that fixes 2 of the problems described there:

2. PR #46347 implemented the mitigation formula incorrectly, which would result in books letting you craft faster when missing a proficiency than when you had the proficiency (if they actually took effect). This caused player confusion, e.g. #46662
When a book provides a proficiency time_factor of 0.5, a malus of 1.5 due to lacking that proficiency should become 1.25 (1 + 0.5 * 0.5) not 0.75 (1.5 * 0.5). The book is supposed to reduce the penalty, not turn it into a bonus.
I confirmed with @I-am-Erk that a 0.1 time_factor in the book JSON is supposed to yield at most 10% mitigation, not 90% or more, so the existing code is also interpreting the value incorrectly.

3. Furthermore, the formula for stacking mitigation effects from multiple books is also wrong and can result in negative proficiency malus.
With 2 books that provide time_factors of 0.1 and 0.5, the formula from #44680 (as implemented in the current code) yields a combined time_factor of 1 - sqrt((1 - 0.1)^2 + (1 - 0.5)^2) = -0.03, which is nonsense.

Fixes #46662

#### Describe the solution

2. Changed the formula in proficiency_time_maluses() as described above.
3. Changed book_proficiency_bonuses::time_factor() and fail_factor() ~~[to simply return the minimum factor provided by any available book]~~ to use the formula suggested by @jbytheway: `1-exp(-sqrt(sum(log(1-t)**2 for t in time_factors)))` which provides diminishing returns similar to the sum-of-squares method used for morale but can never achieve 100% mitigation.

I also removed some dead code that I found while debugging.
What this PR doesn't do is apply mitigations to the actual time taken or chance to fail when crafting something (those changes are in #48650). Only the crafting UI is affected by this PR.

#### Describe alternatives you've considered

~~For 3 I considered using a different formula that would give diminishing returns when stacking books, something like 1/(1/factor1 + 1/factor2 + ...). For 2 books with mitigation factors 0.1 and 0.5 this results in a 0.08 mitigation factor which seems reasonable. But with input factors 0.9 and 0.9 it combines to 0.45 which doesn't seem right.
Alternatively, we could simply multiply the mitigation factors, so 0.1 and 0.5 gives 0.05 and 0.9 with 0.9 gives 0.81. But three 0.5 books yield 0.125 which still feels a bit unrealistic to me because in reality I'd expect the techniques described in the books to overlap significantly.~~

#### Testing

Build 2 game binaries, one without this change ("before") and one with ("after").
Using "before", start a new game. Debug all skills to 10. Debug unlock all recipes. Save.
Bring up the crafting menu and search for "nord".
We're missing the basic carpentry proficiency so without any books we'll see this screen:
<details>
<img src="https://user-images.githubusercontent.com/8000047/116029113-6e491100-a60d-11eb-8d42-6469acb3edaa.png"/>
</details>

Now spawn in "Building for Beginners". This book has proficiency mitigation of 0.5 (time) for carpentry:
```
      { "proficiency": "prof_carpentry_basic", "time_factor": 0.5, "fail_factor": 0.4, "include_prereqs": true },
```
The crafting menu for nord now looks like this:
<details>
<img src="https://user-images.githubusercontent.com/8000047/116029123-73a65b80-a60d-11eb-8d00-a7b0c9aa893e.png"/>
</details>

We see that the lack of proficiency is fully mitigated (shouldn't be, that's not what the book does) and the yellow text claims that it will take 0.8x time (should never be less than 1 when we're missing proficiencies).

Now switch to the "after" binary, load the save, spawn the book and bring up the crafting menu for "nord":
<details>
<img src="https://user-images.githubusercontent.com/8000047/120058700-acb55f80-c001-11eb-84b4-5f448b6e8c17.png"/>
</details>

Now the mitigation reduces the malus to 1.25 and this is consistently shown in both the "Proficiencies Missing" text and the yellow text warning.
Note that the total crafting time is always 22.5 minutes in all cases with or without the book (not being fixed in this PR).

Now let's try stacking books.
Switch to "before", load the save and bring up the crafting menu for "zentai":
<details>
<img src="https://user-images.githubusercontent.com/8000047/116029147-802ab400-a60d-11eb-8996-ca5f25731de0.png"/>
</details>

Now spawn in "Sew What? Clothing!" and "Boston AnimeCon magazine" (time mitigation factors 0.5 and 0.1) and try again.
<details>
<img src="https://user-images.githubusercontent.com/8000047/116029178-8c167600-a60d-11eb-88f5-2efc6211df6a.png"/>
</details>

We see that the yellow text is predicting -0.1x time taken. We've discovered time travel!

Switch to "after", load, spawn the books, try crafting:
<details>
<img src="https://user-images.githubusercontent.com/8000047/120058867-6d3b4300-c002-11eb-8dc7-0c4a2dc6f03d.png"/>
</details>

Much more reasonable, the time multiplier is 1.5x (mostly due to the "Sew What" book, the magazine has negligible effect). It displays as 1.49603 which is ugly, but I can't easily fix that during the string freeze (I can think of one way but it's kind of hacky so I'll leave that for another PR).

I also tested the JSON validation by modifying the first book in tailor.json and changing time_factor and fail_factor on line 8 to illegal values such as -1, 1 and 2. The expected error message was displayed when loading any save, without crashing or freezing the game.

eltank commented:
Question: Will the format change from %g to %.1f in recipe.cpp prevent this PR from going into 0.F due to string freeze? I could revert that part if yes (and submit a separate PR for it). The change is needed because %g can output a lot of decimals sometimes, though that's much less likely after the formula changes in this PR.

actual-nh commented:
> Question: Will the format change from %g to %.1f in recipe.cpp prevent this PR from going into 0.F due to string freeze? I could revert that part if yes (and submit a separate PR for it). The change is needed because %g can output a lot of decimals sometimes, though that's much less likely after the formula changes in this PR.

Ping: @BrettDong?

BrettDong commented:
Yes, this PR will not be subject to String Freeze if no string is modified.

eltank commented:
This PR does change a translatable string in recipe.cpp:

```
        return string_format( _( "<color_%s>%s</color> (<color_%s>%gx\u00a0failure</color>%s)" ),
```

Replacing %g with %.1f. Asking if the translation system is smart enough to not require re-translation when something like that changes.

BrettDong commented:
Transifex shows similar previously translated strings but doesn't automatically apply them. Translators will have to manually fill in suggestions and edit `%g` to `%.1f`.

eltank commented:
Thanks. In that case it sounds like I shouldn't make the string change in this PR (which is targeting 0.F) because it will create extra work for translators.

eltank commented:
Could someone review this change? Please let me know if there's something I should do to make that easier.

esotericist commented:
apologies, we're not intending to ignore this, just there's been quite a lot going on that's gotten in the way of looking into things like this.

as for me personally, i'm afraid i don't quite have the c++ knowledge to review the code changes here.

I-am-Erk commented:
At a brief glance this looks promising. I will begin the arduous task of figuring out what I'm looking at.

I-am-Erk commented:
OK. This looks great. Only one suggestion.

The intent with the original code was to make the mitigation multipliers a sum of squares, so if you have a single 0.1 time factor book, it gives you 0.1 mitigation `sqrt(0.1^2)`. If you have two, it gives you 0.14 mitigation `sqrt(0.1^2+0.1^2)`. Looking at your code, I think this is doable but I am not in a position to post a code suggestion from my phone.

I-am-Erk reviewed (changes requested):
In addition to changes above, also, this could use a comment or two so we don't get a repeat of the problem that led to this fix being needed.

eltank commented:
@I-am-Erk Thanks for taking a look!
I'm so sorry, but the rebase force-push seems to have eaten your code suggestions (maybe I should use merge rebase instead?). I should have looked at them before rebasing. Could you re-add them? I promise to not repeat the mistake.

Regarding the sum-of-squares thing. First let's make sure I am interpreting the mitigation factors correctly.
In the case of "Boston AnimeCon magazine" which has `time_factor: 0.1` and `fail_factor: 0.1` for 3 proficiencies, is this book (by itself) supposed to reduce your crafting malus (additional time spent due to lacking a proficiency) by 10% or 90%? In other words, is this supposed to be an amazing aid for crafting or just a slight aid? In the current implementation it is an amazing aid, but I'm not sure if that matches your intentions.
I think I've figured out how to fix the formula in order to provide sum-of-squares diminishing returns, but I need an answer to that question first.

I-am-Erk commented:
> 
> 
> @I-am-Erk Thanks for taking a look!
> I'm so sorry, but the rebase force-push seems to have eaten your code suggestions (maybe I should use merge rebase instead?). I should have looked at them before rebasing. Could you re-add them? I promise to not repeat the mistake.
> 
> Regarding the sum-of-squares thing. First let's make sure I am interpreting the mitigation factors correctly.
> In the case of "Boston AnimeCon magazine" which has `time_factor: 0.1` and `fail_factor: 0.1` for 3 proficiencies, is this book (by itself) supposed to reduce your crafting malus (additional time spent due to lacking a proficiency) by 10% or 90%? In other words, is this supposed to be an amazing aid for crafting or just a slight aid? In the current implementation it is an amazing aid, but I'm not sure if that matches your intentions.
> I think I've figured out how to fix the formula in order to provide sum-of-squares diminishing returns, but I need an answer to that question first.

Gotcha. I had not yet made code suggestions, I was on my phone. I am now home and will be around the computer on and off for a few hours, so once we're done this next bit I'll give some better direct review. With any luck we can merge this today.

`time_factor: 0.1` means it should reduce the current time_multiplier by 10%, so it should be a low quality aid.

In other words the final formula for total `time_multiplier` should be:
`time_multiplier_final = time_multiplier_base * (1 - sqrt( SUM ( time_factor^2 ) ) )`

Kelenius commented:
> In other words the final formula for total `time_multiplier`, figuring in only the book effect, should be:
> `time_multiplier_final = time_multiplier_base * (1 - sqrt( SUM ( time_factor^2 ) ) )`

I think that would make crafting faster without a proficiency with high mitigation.

I-am-Erk commented:
> 
> 
> > In other words the final formula for total `time_multiplier`, figuring in only the book effect, should be:
> > `time_multiplier_final = time_multiplier_base * (1 - sqrt( SUM ( time_factor^2 ) ) )`
> 
> I think that would make crafting faster without a proficiency with high mitigation.

We should have an error trap that ensures time_multiplier_final is never lower than 1, but it is a diminishing returns formula, so unless someone starts putting in an awful lot of high factor items, it should be difficult to get that high.

Kelenius commented:
Nah, the problem is that if you start with, e.g. x2 (2) time multiplier and then put 75% (0.75) mitigation on that, you end up with 50% (2 * (1 - 0.75) = 0.5) crafting time. The actual formula should be
`(time_multiplier_base - 1) * (1 - sqrt( SUM ( time_factor^2 )) + 1`
Which would give the expected 125% crafting time for that example.

I-am-Erk commented:
Ah, I see what you're saying. That's not exactly how I intended it, but I think that possibly produces more of what people would intuitively expect.

eltank commented:
I think what Kelenius wrote is exactly what I ended up implementing. I just need to test a bit before pushing the update to github.

eltank commented:
Ok, the code is updated. I also added a couple of comments to proficiency.h and clamped the total mitigation factor so it never exceeds 1.
Should I also modify `book_proficiency_bonus::deserialize` to make sure mitigation factors in JSON are always in range [0,1]?

eltank commented:
Note that if you manage to find 4 different books that mitigate the proficiency maluses for one prof by 0.5 each you will completely eliminate any penalties due to missing that proficiency. I don't think this situation can occur with the current set of books in the base game, but just wanted to point out that this is a potential flaw in the sum-of-squares algorithm and to check that it's ok.

actual-nh commented:
> Note that if you manage to find 4 different books that mitigate the proficiency maluses for one prof by 0.5 each you will completely eliminate any penalties due to missing that proficiency. I don't think this situation can occur with the current set of books in the base game, but just wanted to point out that this is a potential flaw in the sum-of-squares algorithm and to check that it's ok.

Personally, I would only allow complete elimination of _time_ effects for someone with FAST_READER or similar (and also assuming prior reading of said books so it's possible to look stuff up fast in them). Failure effects are another matter.

eltank commented:
> Personally, I would only allow complete elimination of _time_ effects for someone with FAST_READER or similar (and also assuming prior reading of said books so it's possible to look stuff up fast in them). Failure effects are another matter.

While I agree that it makes sense for stuff like FAST_READER, GOOD_MEMORY and enhanced memory banks to play a role in one's ability to apply knowledge from books, could we agree that it's out of scope for this PR? That's a non-trivial change that would need some careful balancing.

Kelenius commented:
Come to think of it, instead of `(1 - sqrt( SUM ( time_factor^2 ))`, why not use `1 - ( MULTIPLY ( 1 - time_factor ) )`? That would give more organic numbers, and would naturally be between 0 and 1.

For example, having 90% + 90% would give 99% mitigation.

eltank commented:
> Come to think of it, instead of `(1 - sqrt( SUM ( time_factor^2 ))`, why not use `1 - ( MULTIPLY ( 1 - time_factor ) )`? That would give more organic numbers, and would naturally be between 0 and 1.
> 
> For example, having 90% + 90% would give 99% mitigation.

I don't mind making that change, as long as @I-am-Erk is ok with it. It does solve the "achieving full mitigation from stacking partial mitigations" issue, but makes stacking small mitigations more effective (e.g. two 10% books give 19% mitigation instead of 14% with sum-of-squares)

actual-nh commented:
> > Personally, I would only allow complete elimination of _time_ effects for someone with FAST_READER or similar (and also assuming prior reading of said books so it's possible to look stuff up fast in them). Failure effects are another matter.
> 
> While I agree that it makes sense for stuff like FAST_READER, GOOD_MEMORY and enhanced memory banks to play a role in one's ability to apply knowledge from books, could we agree that it's out of scope for this PR? That's a non-trivial change that would need some careful balancing.

Sorry - I was unclear. I would simply try to avoid it being possible to eliminate time penalties, at least in this PR. Additional complexities like the above are _definitely_ for the future.

jbytheway commented:
If you want the sum-of-squares feature and the converging-to-1 feature then you can combine both using a formula along the lines of (using Python notation):
```
1-exp(-sqrt(sum(log(1-t)**2 for t in time_factors)))
```
* For any individual factor that's the identity function.
* Adding any additional factor will always make at least some small improvement.
* For two factors of 0.1 it yields 0.138
* For two factors of 0.5 it yields 0.625
* For two factors of 0.9 it yields 0.961
* For 0.1 and 0.9 it yields 0.9002 (so the 0.1 barely makes any difference).

Those seem like reasonable results at first glance.

(Note that the choice of base of logarithm is irrelevant, so long as the exponential function uses the same base)

I-am-Erk commented:
> > Come to think of it, instead of `(1 - sqrt( SUM ( time_factor^2 ))`, why not use `1 - ( MULTIPLY ( 1 - time_factor ) )`? That would give more organic numbers, and would naturally be between 0 and 1.
> > For example, having 90% + 90% would give 99% mitigation.
> 
> I don't mind making that change, as long as @I-am-Erk is ok with it. It does solve the "achieving full mitigation from stacking partial mitigations" issue, but makes stacking small mitigations more effective (e.g. two 10% books give 19% mitigation instead of 14% with sum-of-squares)

The goal of sum of squares is precisely to mitigate the effect where you've got a really good textbook dropping your penalties by say 50%, and then ten random goofy magazines that somehow are more effective than a major reference book. I would like to keep it.

eltank commented:
Thanks @jbytheway, that's exactly the kind of formula I wish I'd been able to come up with when considering alternatives.

With 4 different books granting 0.5 mitigation you'd end up with 0.75 total mitigation instead of 1.0. 
With 10 books granting 0.1 mitigation you'd end up with 0.28 total mitigation.
These numbers seem reasonable to me. 
Note that in practice we don't actually have that many proficiency-mitigating books for any one proficiency, so this is more of a future-proofing argument.

The caveat is that a book with 100% mitigation factor (1.0) breaks the formula. Such a book should be "illegal" and I'll add some JSON validation to ensure that. It'll probably take me an hour or two before this is ready.

jbytheway commented:
> The caveat is that a book with 100% mitigation factor (1.0) breaks the formula. Such a book should be "illegal" and I'll add some JSON validation to ensure that. It'll probably take me an hour or two before this is ready.

Are you sure that a mitigation of 1 breaks the formula? If have expected `log(0)` (when using the C library `log` function to return minus infinity, and the rest of the formula should work out fine, but I'm not in a position to test it right now.

eltank commented:
> > The caveat is that a book with 100% mitigation factor (1.0) breaks the formula. Such a book should be "illegal" and I'll add some JSON validation to ensure that. It'll probably take me an hour or two before this is ready.
> 
> Are you sure that a mitigation of 1 breaks the formula? If have expected `log(0)` (when using the C library `log` function) to return minus infinity, and the rest of the formula should work out fine, but I'm not in a position to test it right now.

1. Regardless of whether it breaks the formula or not, I believe such a book is unrealistic.
2. Are you sure that floating point operations on infinite numbers are well-defined in whichever C++ standard is being used for CDDA releases? This feels like one of those "better safe than sorry" situations.

eltank commented:
The code is ready to review. I changed the mitigation stacking algorithm using jbytheway's formula and added some JSON validation checks. I added a few debugmsg strings, but they are non-translatable and should not be affected by the string freeze AFAICT.
I had to rebase again in order to pick up #49055, the fix for #49050, which kept crashing the game when I was testing.

It'll take me some time to update the PR description to reflect the most recent changes, but reviewers should not be blocked by that.

jbytheway commented on src/proficiency.cpp:
```diff
@@ -383,19 +383,22 @@ void book_proficiency_bonus::deserialize( JsonIn &jsin )
-        debugmsg( "ERROR: Tried to add two book proficiency bonuses with different ids" );
-        // can't add them together unless the ids are the same
-        return *this;
+    if( fail_factor < 0 ) {
+        debugmsg( "fail_factor for proficiency %s must not be <0, value: %f", id.str(), fail_factor );
+        fail_factor = 0;
+    }
+    if( fail_factor >= 1 ) {
+        debugmsg( "fail_factor for proficiency %s must not be >=1, value: %f", id.str(), fail_factor );
+        fail_factor = 0.999f;
+    }
+    if( time_factor < 0 ) {
+        debugmsg( "time_factor for proficiency %s must not be <0, value: %f", id.str(), time_factor );
+        time_factor = 0;
+    }
+    if( time_factor >= 1 ) {
+        debugmsg( "time_factor for proficiency %s must not be >=1, value: %f", id.str(), time_factor );
+        time_factor = 0.999f;
     }
```
Consider using `jo.throw_error` for these instead of `debugmsg` so that they provide context as to where the error is.  (The disadvantage is that they will then become fatal errors and prevent the game starting, but I don't think that's too big a deal so long as we don't know of any examples that would trigger these errors).

jbytheway commented:
> [This C++ reference page for log](https://www.cplusplus.com/reference/cmath/log/) claims that:
> 
> > If x is zero, it may cause a pole error (depending on the library implementation).

FWIW, [another C++ reference page for log](https://en.cppreference.com/w/cpp/numeric/math/log) guarantees a return value of -infinity.

You're right that the C standard doesn't specify, but my understanding is that any platform adhering to [IEEE 754](https://en.wikipedia.org/wiki/IEEE_754) will behave this way.  I'm not aware of any non-IEEE 754 platforms, and I strongly suspect other parts of the codebase would break on such platforms too.

But if you want to forbid such values on the grounds that they shouldn't exist anyway, then that's fine by me.

I-am-Erk commented:
I think it's best to just forbid the values to prevent wacky fun from occuring.

The math and code here look good to me. I think just change the error message type so that it tosses a fatal error and prevents bad JSON, and then we are good

eltank commented:
Thanks for the tip about throw_error. It makes the code nicer and the error message much more informative.
![json-error](https://user-images.githubusercontent.com/8000047/120084120-ac10dd80-c082-11eb-8547-9d6ee9bc6741.png)

Note that it's not actually fatal. You can resume the game, but due to the json object for the book not getting loaded you'll get more errors about missing items and recipes.

I-am-Erk commented:
All right, well, when I was given merge permissions I said I wouldn't merge significant C++ changes, but I'm a loose cannon and a maverick who should hand in my badge and gun.

eltank commented:
@I-am-Erk Thanks! As a reminder, this PR on its own doesn't change the way crafting works, only what gets shown in the UI. That parent PR needs to be merged as well in order to "fix" book mitigations.

@jbytheway I'd be very grateful if someone with your level of C++ expertise could review that other PR since there's some tricky const/mutable/const_cast shenanigans in there.

I-am-Erk commented:
Yep, that's on my radar for today now that the infrastructure is in.
